### PR TITLE
Improve type resolution on binary expressions involving integer types

### LINKED ISSUE #105: Cardinal to Integer cast marked as redundant cast
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.0.0

### SonarQube version

10.3.0.82913 Community Edition

### Issue description

Following code in Spring.HashTable.pas (line 732) is being marked with `community-delphi:RedundantCast`:

`        8: hashCode := Integer(PInt64Rec(@key).Cardinals[0] xor PInt64Rec(@key).Cardinals[1]) and not RemovedFlag;`

The cast of Cardinal to Integer is there for a reason: to suppress `W1071 Implicit integer cast with potential data loss from 'unsigned 32-bit integer' to 'Integer'`

### Steps to reproduce

Running SonarDelphi on Spring4D 2.0.0

### Minimal Delphi code exhibiting the issue

_No response_

### PR DESCRIPTION
**Pull request by Cirras**
Based on reverse-engineered data from @zaneduffield (see https://github.com/integrated-application-development/sonar-delphi/issues/107#issuecomment-1830841874), this PR significantly improves the type resolution of binary expressions where the operands are integer types.

Some notable changes:

* When calculating the ordinal distance between 2 integer types, the distance is now only calculated between the `max` values.
* The signage of the operand types is taken into consideration.
* Comparisons between signed and unsigned integer types are now modeled more accurately.

Fixes #105

### CODE REVIEW

File: delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+  @CsvSource(textBlock = OVERLOAD_RESOLUTION_TABLE)
+  @ParameterizedTest(name = "{0} + {1} = {2}")
+  void testAddOperatorOverloadResolution(
+      @AggregateWith(ArithmeticDataAggregator.class) ExpressionData expressionData) {
+    assertResolved(expressionData, BinaryOperator.ADD);
+  }
+
+  @CsvSource(textBlock = OVERLOAD_RESOLUTION_TABLE)
+  @ParameterizedTest(name = "{0} - {1} = {3}")
```
Comment (zaneduffield) [inline delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java:214]: Should be `= {2}`

File: delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  void testSubtractOperatorOverloadResolution(
+      @AggregateWith(ArithmeticDataAggregator.class) ExpressionData expressionData) {
+    assertResolved(expressionData, BinaryOperator.SUBTRACT);
+  }
+
+  @CsvSource(textBlock = OVERLOAD_RESOLUTION_TABLE)
+  @ParameterizedTest(name = "{0} * {1} = {3}")
```
Comment (zaneduffield) [inline delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java:221]: Should be `= {2}`

File: delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  void testMultiplyOperatorOverloadResolution(
+      @AggregateWith(ArithmeticDataAggregator.class) ExpressionData expressionData) {
+    assertResolved(expressionData, BinaryOperator.MULTIPLY);
+  }
+
+  @CsvSource(textBlock = OVERLOAD_RESOLUTION_TABLE)
+  @ParameterizedTest(name = "{0} div {1} = {3}")
```
Comment (zaneduffield) [inline delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java:228]: Should be `= {2}`

File: delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+  void testDivOperatorOverloadResolution(
+      @AggregateWith(ArithmeticDataAggregator.class) ExpressionData expressionData) {
+    assertResolved(expressionData, BinaryOperator.DIV);
+  }
+
+  @CsvSource(textBlock = OVERLOAD_RESOLUTION_TABLE)
+  @ParameterizedTest(name = "{0} mod {1} = {3}")
```
Comment (zaneduffield) [inline delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/OperatorOverloadResolutionTest.java:235]: Should be `= {2}`

**Review by zaneduffield — CHANGES_REQUESTED** [review]
Looks really good! My only review points are to do with comments and admin.

File: delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/resolve/TypeComparer.java [hunk 5]
```diff
@@ -578,9 +580,17 @@ private static EqualityType compareArrayConstructorToSet(
...
-    var subEquality = compare(getLargestType(from.elementTypes()), to.elementType());
+    EqualityType subEquality;
+    Type fromElementType = getLargestType(from.elementTypes());
+
+    if (isConvertibleToSetElementType(fromElementType, to.elementType())) {
+      subEquality = EQUAL;
+    } else {
+      // this should lose to the array constructor -> open array conversions,
+      // but it might happen that the end of the convert levels is reached
+      subEquality = compare(fromElementType, to.elementType());
```
[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/resolve/TypeComparer.java:591]: Still relevant, but it should be moved down - good eye.

File: delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/resolve/TypeComparer.java [hunk 6]
```diff
@@ -596,6 +606,35 @@ private static EqualityType compareArrayConstructorToSet(
     }
   }
 
+  private static boolean isConvertibleToSetElementType(Type from, Type to) {
+    // HACK:
```
[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/main/java/au/com/integradev/delphi/symbol/resolve/TypeComparer.java:610]: Created an issue for constant expressions: https://github.com/integrated-application-development/sonar-delphi/issues/116

I'll add a reference to the issue in the comment.

**Review by zaneduffield — APPROVED** [review]

**Pull request merged**